library/psa_crypto_mac.c

```diff
@@ -382,22 +382,36 @@ psa_status_t mbedtls_psa_mac_update(
 
 static psa_status_t psa_mac_finish_internal(
     mbedtls_psa_mac_operation_t *operation,
-    uint8_t *mac, size_t mac_size)
+    uint8_t *mac, size_t mac_size, size_t *mac_length)
 {
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    size_t actual_mac_length = 0;
 #if defined(MBEDTLS_PSA_BUILTIN_ALG_CMAC)
     if (PSA_ALG_FULL_LENGTH_MAC(operation->alg) == PSA_ALG_CMAC) {
         uint8_t tmp[PSA_BLOCK_CIPHER_BLOCK_MAX_SIZE];
+        actual_mac_length = PSA_MAC_LENGTH(PSA_KEY_TYPE_DES, 0, operation->alg);
```

Comment:
This is a workaround just so that I can get the code out there. But this needs a proper fix.
Problem for me is I cannot access the key type easily. I welcome suggestions.

HMAC is easier to handle, but I want first to have the cmac solution to see how I can unify the two

Reply:
Transferring from our Slack discussion where we were cut off by the server failure: it is possible for `operation->alg` to encode the hash length. In the setup function, we can do
```
size_t length = PSA_MAC_LENGTH(alg, psa_get_key_type(attributes));
operation->alg = PSA_ALG_TRUNCATED_MAC(length, alg);
```
Then `finish` can always extract the desired output length from `operation->alg`.

Contrary to what I initially wrote, this doesn't even require any extra guarantees on the interface between the core and the driver, we can do this entirely inside the driver code.

Reply:
Hopefully I've done what you suggested in the way you envisioned it